torch/utils/hooks.py

```diff
@@ -58,3 +61,110 @@ def warn_if_has_hooks(tensor):
                               "serialized.  If this is expected, you can "
                               "decorate the function with @torch.utils.hooks.unserializable_hook "
                               "to suppress this warning".format(repr(hook)))
+
+class BackwardHook(object):
+    def __init__(self, module, user_hooks):
+        self.user_hooks = user_hooks
+        self.module = module
+
+        self.grad_outputs = None
+        self.n_outputs = -1
+        self.output_tensors_index = None
+        self.n_inputs = -1
+        self.input_tensors_index = None
+
+    def _pack_with_none(self, indices, values, size):
+        res = [None] * size
+        for idx, val in zip(indices, values):
+            res[idx] = val
+
+        return tuple(res)
+
+    def _unpack_none(self, indices, values):
+        res = []
+        for idx in indices:
+            res.append(values[idx])
+
+        return tuple(res)
+
+    def _set_user_hook(self, grad_fn, user_hook):
+        @functools.wraps(user_hook)
+        def hook(grad_input, _):
+            if self.grad_outputs is None:
+                raise RuntimeError("Module backward hook for grad input is called "
```

Comment:
I'm having a hard time with this error message. Could it be more direct in explaining what the issue is and how to resolve it?

Reply:
The thing is that I can't think of a simple case where this would happen.
The user would need to have an input that requires grad + and output that does not (or at least is not used to compute the loss) + they extract a Tensor that requires grad based on the input and use that to compute the loss.
I am not sure how to capture this easily though.

Reply:
Maybe you could just say something like: "Missing grad outputs. (suggested remedy goes here)." Or, since it's rare that this will occur, maybe it doesn't need to be improved.

Reply:
Is the following understandable: "This happens because the gradient in your nn.Module flow to the input without passing by the output. Make sure that the output depends on the input and that the loss is computed based on the output."

Reply:
> Is the following understandable: "This happens because the gradient in your nn.Module flow to the input without passing by the output. Make sure that the output depends on the input and that the loss is computed based on the output."

Makes a lot of sense. Two minor corrections: 

"This happens because the gradient in your nn.Module flows to the input without passing through the output. Make sure that the output depends on the input and that the loss is computed based on the output."